app/src/main/java/de/westnordost/streetcomplete/data/osm/mapdata/MapDataCache.kt

```diff
@@ -0,0 +1,470 @@
+
+    /**
+     * Gets all relations for the node with the given [id] from cache. If the list of relations is
+     * not known, or any relation is missing in cache, [fetch] is called and the result cached.
+     */
+    fun getRelationsForNode(id: Long, fetch: (Long) -> List<Relation>) =
+        getRelationsForElement(ElementType.NODE, id) { fetch(id) }
+
+    /**
+     * Gets all relations for way with the given [id] from cache. If the list of relations is not
+     * known, or any relation is missing in cache, [fetch] is called and the result cached.
+     */
+    fun getRelationsForWay(id: Long, fetch: (Long) -> List<Relation>) =
+        getRelationsForElement(ElementType.WAY, id) { fetch(id) }
+
+    /**
+     * Gets all relations for way with the given [id] from cache. If the list of relations is not
+     * known, or any relation is missing in cache, [fetch] is called and the result cached.
+     */
+    fun getRelationsForRelation(id: Long, fetch: (Long) -> List<Relation>) =
+        getRelationsForElement(ElementType.RELATION, id) { fetch(id) }
+
+    private fun getRelationsForElement(
+        type: ElementType,
+        id: Long,
+        fetch: () -> List<Relation>
+    ): List<Relation> = synchronized(this) {
+        val relationIds = relationIdsByElementKeyCache.getOrPut(ElementKey(type, id)) {
+            val relations = fetch()
+            for (relation in relations) { relationCache[relation.id] = relation }
+            relations.map { it.id }.toMutableList()
+        }
+        return relationIds.mapNotNull { relationCache[it] }
+    }
+
+    /**
+     * Gets all elements and geometries inside [bbox]. This returns all nodes, all ways containing
+     * at least one of the nodes, and all relations containing at least one of the ways or nodes,
+     * and their geometries.
+     * If data is not cached, tiles containing the [bbox] are fetched from database and cached.
+     */
+    fun getMapDataWithGeometry(bbox: BoundingBox): MutableMapDataWithGeometry = synchronized(this) {
+        val requiredTiles = bbox.enclosingTilesRect(tileZoom).asTilePosSequence().toList()
+        val cachedTiles = spatialCache.getTiles()
+        val tilesToFetch = requiredTiles.filterNot { it in cachedTiles }
+        val tilesRectToFetch = tilesToFetch.minTileRect()
+
+        val result = MutableMapDataWithGeometry()
+        result.boundingBox = bbox
+        val nodes: Collection<Node>
+        if (tilesRectToFetch != null) {
+            // fetch needed data
+            val fetchBBox = tilesRectToFetch.asBoundingBox(tileZoom)
+            val (elements, geometries) = fetchMapData(fetchBBox)
+
+            // get nodes from spatial cache
+            // this may not contain all nodes, but tiles that were cached initially might
+            // get dropped when the caches are updated
+            // duplicate fetch might be unnecessary in many cases, but it's very fast anyway
+
+            nodes = HashSet<Node>(spatialCache.get(bbox))
```

Comment:
Hmmm :-/ , so `update` trims? That means that the cache is trimmed twice no? Once in `update` and once at the end of this method